kdc/httpkadmind.c

```diff
@@ -0,0 +1,2025 @@
+            ret = get_kadm_handle(r->context, &r->kadm_handle, 1);
+        } else if (ret == 0 && local_read_only) {
+            ret = KADM5_READ_ONLY;
+        }
+
+        /* Use requested enctypes or same ones as princ already had keys for */
+        if (ret == 0 && kstuple == NULL)
+            ret = make_kstuple(r->context, &princ, &kstuple, &nkstuple);
+
+        /* Set new keys */
+        if (ret == 0)
+            ret = kadm5_randkey_principal_3(r->kadm_handle, p, keepold,
+                                            nkstuple, kstuple, &k, &n_k);
+        refetch = 1;
+        for (i = 0; n_k > 0 && i < n_k; i++)
+            krb5_free_keyblock_contents(r->context, &k[i]);
+        free(kstuple);
+        free(k);
+    }
+
+    if (ret == 0 && refetch) {
+        /* Refetch changed principal */
+        if (freeit)
+            kadm5_free_principal_ent(r->kadm_handle, &princ);
+        freeit = 0;
+        ret = kadm5_get_principal(r->kadm_handle, p, &princ, mask);
+        if (ret == 0)
+            freeit = 1;
+    }
+
+    if (ret == 0)
+        ret = write_keytab(r, &princ, pname);
+    if (freeit)
+        kadm5_free_principal_ent(r->kadm_handle, &princ);
+    krb5_free_principal(r->context, p);
+    return ret;
+}
+
+static krb5_error_code
+get_keysN(kadmin_request_desc r)
+{
+    krb5_error_code ret;
+    size_t nhosts;
+    size_t nsvcs;
+    size_t nspns;
+    size_t i, k;
+
+    /* Parses and validates the request, then checks authorization */
+    ret = authorize_req(r);
+    if (ret)
+        return ret; /* authorize_req() calls bad_req() */
+
+    nhosts = heim_array_get_length(r->hostnames);
+    nsvcs = heim_array_get_length(r->service_names);
+    nspns = heim_array_get_length(r->spns);
+    if (!nhosts && !nsvcs && !nspns) {
+        krb5_set_error_message(r->context, ret = EINVAL,
+                               "No service principals requested");
+        return ret;
+    }
+
+    if (nhosts && !nsvcs) {
+        heim_string_t s;
+
+        if ((s = heim_string_create("HTTP")) == NULL)
+            ret = krb5_enomem(r->context);
+        if (ret == 0)
+            ret = heim_array_append_value(r->service_names, s);
+        heim_release(s);
+        nsvcs = 1;
+    }
+
+    ret = make_keytab(r);
+
+    for (i = 0; ret == 0 && i < nsvcs; i++) {
+        const char *svc =
+            heim_string_get_utf8(
+                heim_array_get_value(r->service_names, i));
+
+        for (k = 0; ret == 0 && k < nhosts; k++) {
+            const char *hostname =
+                heim_string_get_utf8(
+                    heim_array_get_value(r->hostnames, k));
+            char *spn = NULL;
+
+            if (asprintf(&spn, "%s/%s", svc, hostname) == -1 ||
+                spn == NULL)
+                ret = krb5_enomem(r->context);
+            if (ret == 0)
+                ret = get_keys1(r, spn);
+            free(spn);
+        }
+    }
+    for (i = 0; ret == 0 && i < nspns; i++) {
+        ret = get_keys1(r,
```

Comment:
Here or somewhere, maybe the authorizer function, or the q-param iterator, check the service name is allowable.